### ISSUE TITLE
Support IsolationLevels

### ISSUE THREAD

**Issue by syun64** [apache/iceberg-python#819]
### Feature Request / Improvement

Support Isolation Levels from specified snapshot ID 

https://iceberg.apache.org/docs/latest/spark-configuration/#write-options

**Comment by jqin61** [comment]
Hi I am interested in working on it!

**Comment by syun64** [comment]
Some relevant links to the Java implementation

- validateNewDataFiles flag -> MergingSnapshotProducer: [validateAddedDataFiles](https://github.com/apache/iceberg/blob/c7d3ef44367c46875f0367d312372fb2f246ae73/core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java#L299-L317)
- validateNewDeletes flag -> MergingSnapshotProducer: [validateDeletedDataFiles](https://github.com/apache/iceberg/blob/c7d3ef44367c46875f0367d312372fb2f246ae73/core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java#L572-L590)
- [validateNoNewDeleteFiles](http://validatenonewdeletefiles/) -> [failMissingDeletePaths](https://github.com/apache/iceberg/blob/c7d3ef44367c46875f0367d312372fb2f246ae73/core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java#L104) flag -> ManifestFilterManager: [validateRequiredDeletes](https://github.com/apache/iceberg/blob/c7d3ef44367c46875f0367d312372fb2f246ae73/core/src/main/java/org/apache/iceberg/ManifestFilterManager.java#L234-L242)

**Comment by guptaakashdeep** [comment]
Hey @sungwy I would like to contribute by working on these. 

Is there any of these that I can pick and starts looking into it like any of the initial validation implementation ?

**Comment by sungwy** [comment]
@guptaakashdeep yes, I don't think there's a particular order we should implement these with, so please feel free to assign yourself to the one you find most interesting!

Sung

**Comment by sungwy** [comment]
I think we could create a new module as `pyiceberg.table.update.validate.py` and add these validation checks there. What do you think @guptaakashdeep ?

**Comment by guptaakashdeep** [comment]
Sounds good @sungwy !!

**Comment by jayceslesar** [comment]
@guptaakashdeep @sungwy see https://github.com/apache/iceberg-python/pull/1935 which should be the building blocks needed to crank out the 4 Sub-issues

**Comment by jayceslesar** [comment]
Also going to crank out a manifest group implementation today

**Comment by cnatsis** [comment]
Is there any update on this feature?

**Comment by dyami0123** [comment]
Also curious if there is any movement on this PR. I currently have some workarounds for concurrent writes implemented but they are very inefficient.

**Comment by lawofcycles** [comment]
@sungwy @gabeiglio
Hi, I'm very interested in helping move this forward. Proper optimistic concurrency support is critical for production use cases with concurrent writers, and I'd like to contribute to getting this completed.

Before jumping in, I wanted to make sure I understand the current state correctly

- #1928 (`validateDeletedDataFiles`) and #1929 (`validateAddedDataFiles`) are completed and merged.
- #1930 (`validateNoNewDeleteFiles`) and #1931 (`validateNoNewDeletesForDataFiles`) were closed by the stale bot, but the implementation work is still outstanding. PR #2176 addresses #1930 and remains open.

I'd love to help with #1930 and #1931, and eventually the isolation level integration. Could you let me know if there's a plan
already in place, or if it would be okay for me to pick these up?

**Comment by gabeiglio** [comment]
Hi @lawofcycles thanks you for bringing this back to my inbox! I have the changes ready to submit for those two issues! The main blocker is this [PR](https://github.com/apache/iceberg-python/pull/2255) which adds the needed DeleteFileIndex implementation. I see that some related PR was merged implementing DeleteFileIndex which I think with some additions can be enough to unblock me. I will try to put the PR up for review tomorrow.

And then we can plan how to add the validation logic in _OverwriteFIles :)

**Comment by lawofcycles** [comment]
Thanks for the quick response, @gabeiglio! I see you've already opened PR #3049, that's great.

**Comment by dyami-andrews-e3** [comment]
With all of these PR's merged is this issue addressed, or will there need to be a final PR to bring everything together?

**Comment by lawofcycles** [comment]
Now that all four validation functions are merged (#1935, #1938, #2050, #3049), I'd like to work on integrating them into the write path.

I've put together a design proposal in #3319 that covers both the validation integration and commit retry (#269) as a single change, since the two are tightly coupled in Java's `SnapshotProducer.commit()` (the retry loop re-runs validation on each attempt).

Would appreciate feedback on the approach.

cc @sungwy @gabeiglio @kaushiksrini @jayceslesar @Fokko @kevinjqliu

**Comment by lawofcycles** [comment]
I've opened #3320 with the implementation. It integrates the validation functions. Would appreciate review and feedback.

**Comment by sungwy** [comment]
Thanks @lawofcycles - let me find time to take a look over the next few days.
